### PR TITLE
#1431 `OptimizeMojo` caching inside `.eo/optimized`

### PR DESCRIPTION
**Pull request by volodya-lombrozo**
What have been done:
1) Added `optimization` package and `Optimization` interface.
2) Added caching mechanism for `OptimizeMojo` through `OptCache` implementation
3) Also I had to open some utility classes in order to use them in a foreign package

Closes: #1431

### CODE REVIEW

**Comment by volodya-lombrozo** [comment]
@mximp please, take a look.

File: eo-maven-plugin/src/main/java/org/eolang/maven/optimization/OptimizationException.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * SOFTWARE.
+ */
+package org.eolang.maven.optimization;
+
+/**
+ * Exception that can happen during XML optimizations.
+ *
+ * @since 0.28.11
+ */
+public class OptimizationException extends Exception {
```
Comment (mximp) [inline eo-maven-plugin/src/main/java/org/eolang/maven/optimization/OptimizationException.java:31]: @volodya-lombrozo I believe we don't need new specific type of exception. We could use some generic (e.g. `IllegalStateExcpetion`) where needed. Ideally we could create single  checked `EoMavenPluginException` and use it this everywhere.

Comment (volodya-lombrozo) [reply eo-maven-plugin/src/main/java/org/eolang/maven/optimization/OptimizationException.java]: @mximp, fixed.

File: eo-maven-plugin/src/main/java/org/eolang/maven/optimization/Optimization.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+public interface Optimization {
+
+    /**
+     * Optimize XML file.
+     *
+     * @param xml Path to raw XML file.
+     * @return Optimized XML
+     * @throws OptimizationException if optimization can't be performed by any reason.
+     */
+    XML optimize(Path xml) throws OptimizationException;
```
[the remark this answers is not in the document]
Comment (volodya-lombrozo) [reply eo-maven-plugin/src/main/java/org/eolang/maven/optimization/Optimization.java]: @mximp, fixed.

File: eo-maven-plugin/src/main/java/org/eolang/maven/optimization/OptLambda.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+     * The main constructor.
+     *
+     * @param delegate Custom optimization.
+     */
+    public OptLambda(final Func<Path, XML> delegate) {
+        this.delegate = new UncheckedFunc<>(delegate);
+    }
+
+    @Override
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
```
[the remark this answers is not in the document]
Comment (volodya-lombrozo) [reply eo-maven-plugin/src/main/java/org/eolang/maven/optimization/OptLambda.java]: @mximp, fixed.

File: eo-maven-plugin/src/main/java/org/eolang/maven/OptimizeMojo.java [hunk 4]
```diff
@@ -175,15 +187,15 @@ public void exec() throws IOException {
...
+                                    final XML optimized = this.optimization(tojo).optimize(src);
                                     done.incrementAndGet();
                                     if (this.shouldPass(optimized)) {
                                         tojo.set(
                                             AssembleMojo.ATTR_XMIR2,
                                             this.make(optimized, src).toAbsolutePath().toString()
                                         );
                                     }
-                                } catch (final IOException exception) {
+                                } catch (final IOException | OptimizationException exception) {
```
[the remark this answers is not in the document]
Comment (volodya-lombrozo) [reply eo-maven-plugin/src/main/java/org/eolang/maven/OptimizeMojo.java]: @mximp, fixed.

**Comment by volodya-lombrozo** [comment]
@mximp all suggestions were applied. Please, check.

**Comment by volodya-lombrozo** [comment]
@yegor256 merge, please.

File: eo-maven-plugin/src/main/java/org/eolang/maven/optimization/Optimization.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ */
+public interface Optimization {
+
+    /**
+     * Optimize XML file.
+     *
+     * @param xml Path to raw XML file.
+     * @return Optimized XML
+     */
+    XML apply(Path xml);
```
Comment (yegor256) [inline eo-maven-plugin/src/main/java/org/eolang/maven/optimization/Optimization.java:42]: @volodya-lombrozo we have `Function` interface in Java. Maybe it's better to use it?

**Comment by yegor256** [comment]
@volodya-lombrozo this `Optimization` is a functional interface. It's good to have it `extend` `Function` or at least annotate it with `@FunctionalInterface`: https://docs.oracle.com/javase/8/docs/api/java/lang/FunctionalInterface.html

File: eo-maven-plugin/src/main/java/org/eolang/maven/optimization/Optimization.java (hunk 5 above)
Comment (volodya-lombrozo) [reply eo-maven-plugin/src/main/java/org/eolang/maven/optimization/Optimization.java]: @yegor256 fixed.

**Comment by volodya-lombrozo** [comment]
> @volodya-lombrozo this `Optimization` is a functional interface. It's good to have it `extend` `Function` or at least annotate it with `@FunctionalInterface`: https://docs.oracle.com/javase/8/docs/api/java/lang/FunctionalInterface.html

I've used both `@FunctionalInterface` and `extends Function`. Check, please

**Comment by yegor256** [comment]
@rultor merge

**Comment by volodya-lombrozo** [comment]
@yegor256 restart merge, please

**Comment by yegor256** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor merge

**Comment by volodya-lombrozo** [comment]
@yegor256 please, merge

**Comment by mximp** [comment]
@volodya-lombrozo the same test fails in master. Are you able to build locally?

**Comment by volodya-lombrozo** [comment]
> @volodya-lombrozo the same test fails in master. Are you able to build locally?

@mximp please, update master. https://github.com/objectionary/eo/pull/1438

**Comment by yegor256** [comment]
@rultor merge

**Pull request merged**